lib/utils/log/slog_handler.go

```diff
@@ -0,0 +1,386 @@
+	}
+
+	*buf = s.appendAttr(*buf, slog.String(trace.Component, s.component))
+
+	*buf = s.appendAttr(*buf, slog.String(slog.MessageKey, r.Message))
+
+	// Insert preformatted attributes just after built-in ones.
+	*buf = append(*buf, s.preformatted...)
+	if r.NumAttrs() > 0 {
+		*buf = s.appendUnopenedGroups(*buf)
+		r.Attrs(func(a slog.Attr) bool {
+			*buf = s.appendAttr(*buf, a)
+			return true
+		})
+	}
+
+	if r.PC != 0 {
+		fs := runtime.CallersFrames([]uintptr{r.PC})
+		f, _ := fs.Next()
+
+		src := &slog.Source{
+			Function: f.Function,
+			File:     f.File,
+			Line:     f.Line,
+		}
+
+		file, line := getCaller(slog.Attr{Key: slog.SourceKey, Value: slog.AnyValue(src)})
+		*buf = fmt.Appendf(*buf, " %s:%d", file, line)
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	_, err := s.out.Write(*buf)
+	return err
+}
+
+// WithAttrs clones the current handler with the provided attributes
+// added to any existing attributes. The values are preformatted here
+// so that they do not need to be formatted per call to Handle.
+func (s *SlogTextHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	if len(attrs) == 0 {
+		return s
+	}
+	s2 := *s
+	// Force an append to copy the underlying array.
+	pre := slices.Clip(s.preformatted)
+	// Add all groups from WithGroup that haven't already been added.
+	s2.preformatted = s2.appendUnopenedGroups(pre)
+	// Now all groups have been opened.
+	s2.unopenedGroups = nil
+
+	component := s.component
+
+	// Pre-format the attributes.
+	for _, a := range attrs {
+		switch a.Key {
+		case trace.Component:
+			const padding = trace.DefaultComponentPadding
+			component = fmt.Sprintf("[%v]", a.Value.String())
+			component = strings.ToUpper(padMax(component, padding))
+			if component[len(component)-1] != ' ' {
+				component = component[:len(component)-1] + "]"
+			}
+		case trace.ComponentFields:
+			switch fields := a.Value.Any().(type) {
+			case map[string]any:
+				for k, v := range fields {
+					s2.appendAttr(s2.preformatted, slog.Any(k, v))
+				}
+			case logrus.Fields:
+				for k, v := range fields {
+					s2.preformatted = s2.appendAttr(s2.preformatted, slog.Any(k, v))
+				}
+			}
+		default:
+			s2.preformatted = s2.appendAttr(s2.preformatted, a)
+		}
+	}
+	s2.component = component
+	return &s2
+}
+
+func (s *SlogTextHandler) appendUnopenedGroups(buf []byte) []byte {
+	for _, g := range s.unopenedGroups {
+		buf = fmt.Appendf(buf, "%s:", g)
+	}
+	return buf
+}
+
+// WithGroup opens a new group.
+func (s *SlogTextHandler) WithGroup(name string) slog.Handler {
+	if name == "" {
+		return s
+	}
+
+	s2 := *s
+	// Add an unopened group to h2 without modifying h.
+	s2.unopenedGroups = make([]string, len(s.unopenedGroups)+1)
+	copy(s2.unopenedGroups, s.unopenedGroups)
+	s2.unopenedGroups[len(s2.unopenedGroups)-1] = name
+	return &s2
+}
+
+// SlogJSONHandler is a [slog.Handler] that outputs messages in a json
+// format per the config file.
+type SlogJSONHandler struct {
+	handler *slog.JSONHandler
```

Comment:
Super nit/musing: We could embed this here to avoid the need to write out the methods that delegate to `handler`.